sdk/formrecognizer/Azure.AI.FormRecognizer/src/CopyModelOperation.cs

```diff
@@ -86,6 +86,12 @@ public override CustomFormModelInfo Value
         /// <param name="client">The client used to check for completion.</param>
         public CopyModelOperation(string operationId, string targetModelId, FormTrainingClient client)
         {
+            // We can't check if `targetModelId` is empty because it will create a breaking change for
+            // released version 3.0.0 of the SDK which works if value is == `string.Empty`.
+            Argument.AssertNotNull(targetModelId, nameof(targetModelId));
+            Argument.AssertNotNullOrEmpty(operationId, nameof(operationId));
+            Argument.AssertNotNull(client, nameof(client));
```

Comment:
Actually, simply adding parameter validation is already a breaking change, isn't it? We're now throwing from a constructor that couldn't throw exceptions before.

Reply:
True-ish. So even though we were not validating the parameters before, our code was counting on that. As an example:
- if the `operationId` was null then the code will throw `NullReferenceException`. It is [not considered to be a breaking change](https://docs.microsoft.com/en-us/dotnet/core/compatibility/#exceptions) to move from `NullReferenceException` to `ArgumentNullException` .

> `string[] substrs = operationId.Split('/');` 
 
- if the `operationId` was empty this will happen:
> if (substrs.Length < 3)
   {
        throw new ArgumentException($"Invalid '{nameof(operationId)}'. It should be formatted as: '{{modelId}}/copyresults/{{resultId}}'.", nameof(operationId));
    }

Same exception that we will throw on ` Argument.AssertNotNullOrEmpty(operationId, nameof(operationId));`. so not breaking change :)

So overall I did this exercise with every parameter I added validation for. If there was a change that is consider breaking, I added it to: #17383  . This PR shouldn't introduce breaking changes. I will do those in a future PR.

Reply:
Let me know if this is not clear.
Should I add more information in the description or in each parameter about the behavior before and after my change to make sure it is right?

Reply:
I see. I wasn't aware of:

> ALLOWED: Throwing a more specific exception than NotSupportedException, NotImplementedException, NullReferenceException


Very helpful.

Reply:
Yeah, I learned about it with the work in TA moving to codegen. Also, https://aka.ms/fxdg3/breaking#secD_12_1: this is more specific:
> D.12.1 Changing Runtime Error Exceptions to Usage Error Exceptions
If your method throws a NullReferenceException when a parameter is passed as null, or an IndexOutOfRangeException when a parameter is passed as an out-of-bounds index, you can change it to throw the usage error equivalent exception (e.g., ArgumentNullException or ArgumentOutOfRangeException) instead. While some callers could, conceivably, depend on the NullReferenceException or IndexOutOfRangeException, the .NET BCL team believes that the information provided via the usage error exception about which argument value caused the method to fail is worth the risk.
 The .NET BCL team generally considers changing “failed inputs” to “a detailed usage error exception” to be an acceptable breaking change.

Reply:
> Should I add more information in the description or in each parameter about the behavior before and after my change to make sure it is right?

No worries. I got your reasoning.

> if the operationId was null then the code will throw NullReferenceException. It is not considered to be a breaking change to move from NullReferenceException to ArgumentNullException .

But the exceptions will be thrown from different places, right? Previously, we were throwing a `NullReferenceException` when making a service call. Now we are failing in the constructor. Shouldn't it be considered breaking in this case?

Reply:
Could you expand more on what do you mean by making a service call?

Before this change, if `operationId == null` the constructor will throw in [here](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/formrecognizer/Azure.AI.FormRecognizer/src/CopyModelOperation.cs#L99) .

Reply:
You're right about `CopyModelOperation`. However, if we take `RecognizeContentOperation` as an example:

Before validation was added:
```cs
var operation = new RecognizeContentOperation(null, client);
await operation.WaitForCompletionAsync(); // Throws NullReferenceException here
```

After validation was added:
```cs
var operation = new RecognizeContentOperation(null, client); // Throws ArgumentNullException here
await operation.WaitForCompletionAsync();
```

I agree that changing the exception type is allowed in this case, but my concern is that we're not throwing from the same place. From my understanding, we're now throwing from a place that didn't throw before. Isn't this a breaking change?

Reply:
Just realized the whole confusion started because I added the comment in `CopyModelOperation`. Apologies for that 👀

Reply:
Ahh I see. That is a great point that I did not consider.
Let me remove the validation of parameters that will throw in other areas and I will add this to issue #17383 so it can be discussed in an arch board meeting

Reply:
ok. removed

Reply:
I updated the description of the issue too